app/lib/features/chat_ng/widgets/chat_messages.dart

```diff
@@ -1,27 +1,92 @@
-    final messages = ref
-        .watch(chatStateProvider(roomId).select((value) => value.messageList));
-    final animatedListKey = ref.watch(animatedListChatMessagesProvider(roomId));
-
-    return AnimatedList(
-      initialItemCount: messages.length,
-      reverse: false,
-      key: animatedListKey,
-      itemBuilder: (_, index, animation) => ChatEventWidget(
-        roomId: roomId,
-        eventId: messages[index],
-        animation: animation,
-      ),
+  ConsumerState<ChatMessages> createState() => _ChatMessagesConsumerState();
+}
+
+class _ChatMessagesConsumerState extends ConsumerState<ChatMessages> {
+  final ScrollController _scrollController = ScrollController();
+  bool isLoading = false;
+
+  @override
+  void initState() {
+    super.initState();
+    _scrollController.addListener(onScroll);
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      scrollToBottom();
+    });
+  }
+
+  Future<void> onScroll() async {
+    if (isLoading) return;
+
+    // Check if we're near the top of the list
+    if (_scrollController.position.pixels >=
+        _scrollController.position.maxScrollExtent) {
+      setState(() => isLoading = true);
+
+      // Get the notifier to load more messages
+      final notifier = ref.read(chatStateProvider(widget.roomId).notifier);
+      await notifier.loadMore();
+
+      setState(() => isLoading = false);
+    }
+  }
+
+  void scrollToBottom() {
+    if (_scrollController.hasClients) {
+      _scrollController.animateTo(
+        _scrollController.position.minScrollExtent,
+        duration: const Duration(milliseconds: 300),
+        curve: Curves.easeOut,
+      );
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final messages = ref.watch(
+      chatStateProvider(widget.roomId).select((value) => value.messageList),
+    );
+    final animatedListKey =
+        ref.watch(animatedListChatMessagesProvider(widget.roomId));
+    return Column(
+      children: [
+        Padding(
+          padding: const EdgeInsets.symmetric(vertical: 12.0),
+          child: AnimatedContainer(
+            duration: const Duration(milliseconds: 200),
+            height: isLoading ? 14 : 0,
+            width: isLoading ? 14 : 0,
+            child: Center(
+              child: isLoading ? const CircularProgressIndicator() : null,
+            ),
+          ),
+        ),
+        // Messages list takes remaining space
+        Expanded(
+          child: AnimatedList(
+            initialItemCount: messages.length,
+            key: animatedListKey,
+            controller: _scrollController,
+            reverse: true,
+            itemBuilder: (_, index, animation) => Padding(
+              padding: const EdgeInsets.symmetric(vertical: 12),
+              child: ChatEvent(
+                roomId: widget.roomId,
+                eventId: messages[messages.length - 1 - index],
+              ),
+            ),
+          ),
```

Comment:
so we reverse to then in the itemBuilder reverse again? Are we sure this isn't messing with the `animatedListKey`'s idea of the index? What if we deleted or added entries? I feel a bit uneasy about this ... Can I have some tests that show this is working fine?